Update for second Beta with GA criteria for "KEP-3329: Retriable and non-retriable Pod failures for Jobs"

mimowo opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Update graduation criteria for GA

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/3329

<!-- other comments or additional information -->
- Main changes:
  - proposal to modify Kubelet to transition into Failed scheduled Pending pods which are Terminating (and with finalizer)
  - refreshed GA graduation criteria

mimowo commented:
/sig node
as a participating sig + the proposed changes involve kubelet

mimowo commented:
Please review

/assign @alculquicondor @soltysh for sig-apps POV
/assign @bobbypage  @SergeyKanzhelev for sig-node POV

alculquicondor commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -884,19 +891,52 @@ However, as disabling taint-manager is deprecated it is not a concern for this
-non-scheduled terminating pods (after the Alpha changes for this feature, see
-above). We plan to extend the PodGC controller to also move to the failed phase
-(and delete) scheduled pods in the pending state which are terminating.
+in PodGC. Similarly, we for Alpha we modified PodGC to mark as `Failed`
```
```suggestion
in PodGC. Similarly, for Alpha we modified PodGC to mark as `Failed`
```

alculquicondor commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -884,19 +891,52 @@ However, as disabling taint-manager is deprecated it is not a concern for this
+
+A scheduled pod may also get stuck in a Pending phase (for example due to an
+invalid image reference). In that case, as long as the pod is not Terminating it
+could be a transient issue and Kubelet can may progress after retrying. However,
```
```suggestion
could be a transient issue and Kubelet can make progress after retrying. However,
```

alculquicondor commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -884,19 +891,52 @@ However, as disabling taint-manager is deprecated it is not a concern for this
 KEP.
 
 For Alpha, we implemented a fix for this issue by setting the pod phase as `Failed`
```
I think it's better to put this in two bullet points, one describing the fix for scheduled pods when kubelet disappears and one for terminating unscheduled pods.

mimowo replied:
I see, I restructured the content, but have three points. Hope it makes sense. Also, dropped some less relevant parts for brevity, will re add it requested.

alculquicondor commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -884,19 +891,52 @@ However, as disabling taint-manager is deprecated it is not a concern for this
+A scheduled pod may also get stuck in a Pending phase (for example due to an
+invalid image reference). In that case, as long as the pod is not Terminating it
+could be a transient issue and Kubelet can may progress after retrying. However,
+once such a Pod (in a Pending phase) is also Terminating, Kubelet is not going
```
```suggestion
once such a Pod (in a Pending phase) is also terminating, Kubelet is not going
```

just because terminating is not a phase

alculquicondor commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -884,19 +891,52 @@ However, as disabling taint-manager is deprecated it is not a concern for this
+allowing API server to complete deletion. A prototype implementation is prepared
+here: [Mark Pending Terminating pods as Failed by Kubelet](https://github.com/kubernetes/kubernetes/pull/115331).
+
+Note that, the logic could be scoped to only pods with finalizers to save an
+unnecessary API call by Kubelet for the phase transition in case there is no
+finalizer, as in this case the pods are about to be deleted by API server anyway.
+However, we prefer simplicity of the implementation without an additional special
+case if the performance gain is not clear.
```
We already do the same for Pod GC, so no need to justify again.

mimowo replied:
Dropped the section

alculquicondor commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -884,19 +891,52 @@ However, as disabling taint-manager is deprecated it is not a concern for this
+However, we prefer simplicity of the implementation without an additional special
+case if the performance gain is not clear.
+
+One release after the PodDisruptionCondition feature gate graduates to GA
+we plan to simplify Job controller to consider as failed (and count as such)
+pods which are in terminal phase regardless of the fact if the `podFailurePolicy`
+is specified (see [Deprecation](#deprecation)).
```
It has to be 2 releases after the kubelet changes. I would just postpone the GA release to 1.29.

mimowo replied:
Updated `kep.yaml` accordingly.

mimowo commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -884,19 +891,52 @@ However, as disabling taint-manager is deprecated it is not a concern for this
+However, It is important for the pod failure policy to match only against pods
+in terminal phase to avoid race conditions when pod's end state is modified
+(e.g. pod's conditions or exit codes added) while the pod is being matched
+against the policy. In the first iteration of Beta we only require pods to be in
```
Done

alculquicondor reviewed (commented):
/lgtm

SergeyKanzhelev commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -830,8 +833,11 @@ but also when the system is running low on memory. In such scenario there
-Thus, we decide not to annotate the scenarios with the `ResourceExhausted`
-condition. While there are not known issues with detection of the exceeding of
+Thus, we've decided not to annotate the scenarios with the `ResourceExhausted`
+condition in this in this KEP. Handling of exceeded limits might be done as a
```
```suggestion
condition in this KEP. Handling of exceeded limits might be done as a
```

SergeyKanzhelev commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -883,20 +889,43 @@ is disabled [Disconnected node when taint-manager is disabled](#disconnected-nod
-above). We plan to extend the PodGC controller to also move to the failed phase
-(and delete) scheduled pods in the pending state which are terminating.
+A scheduled pod may also get stuck in a `Pending` phase (for example due to an
+invalid image reference). In that case, as long as the pod is not Terminating it
```
will kubelet retry image download for the pod with `restartPolicy=Never`? I simply don't remember off the top of my head

SergeyKanzhelev replied:
will pod admission failure be in the same category of issues

mimowo replied:
> will kubelet retry image download for the pod with `restartPolicy=Never`? I simply don't remember off the top of my head

Yes, Kubelet attempts to retry image download for as long as it either succeeds or the pod is terminating. I've tested this on KIND by blocking traffic with `iptables` and re-opening the traffic after 2min (after 4 failed attempts). Once re-opened the traffic, the image was pulled and the container started to run.

> will pod admission failure be in the same category of issues

No, when admission failure happens the pod is evicted (marked as `Failed` by Kubelet). More specifically, `rejectPod` is called: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2360-L2361, which sets the phase as `Failed`: https://github.com/kubernetes/kubernetes/blob/7b243cef1a81f4a0ac1619f7b9238fcd2bac2b9c/pkg/kubelet/kubelet.go#L2062-L2065

SergeyKanzhelev commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -883,20 +889,43 @@ is disabled [Disconnected node when taint-manager is disabled](#disconnected-nod
+
+In order to ensure that Job controller only considers pods in terminal phase we
+need to make sure pods in the following states are transitioned into `Failed`
+phase:
```
sorry, I do not understand these phases. Can you please expand the description a bit. I'm lost in this terminology - for example scheduled or unscheduled.

mimowo replied:
Thank you for this comment. It can be confusing. Scheduled means bound to a node (with `.spec.nodeName` assigned). Unscheduled is just the negation of that. Updated the KEP to clarify that and explain what terminating means too.

mimowo replied:
@SergeyKanzhelev I have rephrased the section again based on the David's feedback. Please take another look, hope it gets easier to parse.

SergeyKanzhelev commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -883,20 +889,43 @@ is disabled [Disconnected node when taint-manager is disabled](#disconnected-nod
+need to make sure pods in the following states are transitioned into `Failed`
+phase:
+1. a pod is in Running phase and orphaned
+2. a pod is in Pending phase, terminating and unscheduled
```
is oxford comma ommitted here on purpose? This comes to the question above on expanding phases description

mimowo replied:
It wasn’t deliberate, I think it might be simpler to understand if commas are added in between all terms. Updated accordingly.

mimowo replied:
Rephrased to account for the feedback, PTAL.

SergeyKanzhelev commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -1402,19 +1431,11 @@ Below are some examples to consider, in addition to the aforementioned [maturity
 #### GA
 
 - Address reviews and bug reports from Beta users
```
documentation for this feature is very weak. I found one paragraph mentioning it. I'd suggest to make the detailed documentaiton with examples to be a beta 2 requirement

mimowo replied:
Thank you for the comment. I’m happy to discuss the missing stuff and update it.  

For the context, the current documentation for this feature includes:
- https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy (pod failure policy for Jobs with example configuration using the DisruptionTarget condition)
- https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-conditions (pod disruption conditions added when a pod is deleted, briefly mentioning the transition phase by PodGC, I’m thinking about expanding this paragraph with the kubelet changes).
- https://kubernetes.io/docs/tasks/job/pod-failure-policy/ (example of using the pod failure policy to control handling of pod failures in real life scenarios)

First, I’ve update the PR to add a point to clarify the recent changes (for Beta 2). However, the transitions to the Failed phase seem to be technical fixes, so not sure they require two user-facing examples (they don't require any specific handling in the Job's `.spec.podFailurePolicy`, they just make sure that the pod failure policy is matched against pods in terminal phases), but maybe we could expand the tasks second with another one. Do you think of a particular example to add?

bobbypage commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -883,20 +889,51 @@ is disabled [Disconnected node when taint-manager is disabled](#disconnected-nod
+(see Issue: [Job controller should wait for Pods to terminate to match the failure policy](https://github.com/kubernetes/kubernetes/issues/113855)).
+
+For the second iteration of Beta (1.27), we plan to fix the 3rd scenario by
+modifying Kubelet to transition `Pending`, terminating, and scheduled pods into
```
I would clarify here what you mean by pending, terminating and scheduled pods to be a little more precise. If I understand correctly, the proposal here is:

Kubelet should transition pod a phase to failed if:

1) The spec.nodeName is set to the kubelet owning the pod
2) A deletion timestamp is set
3) The pod still has pending phase.

soltysh replied:
I've been thinking of the same when reading the three scenarios above. Maybe it'll be too vocal, but I think in the long run all will appreciate a bit more details in explaining those scenarios.

mimowo replied:
Thank you for the feedback. I've rephased this section to make it easier to follow. PTAL.

soltysh replied:
This looks great - thank you!

bobbypage commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -883,20 +889,51 @@ is disabled [Disconnected node when taint-manager is disabled](#disconnected-nod
+For the second iteration of Beta (1.27), we plan to fix the 3rd scenario by
+modifying Kubelet to transition `Pending`, terminating, and scheduled pods into
+the `Failed` phase, allowing the Job
+controller to count them, match against the (optional) pod failure policy and
+remove their Job finalizers, allowing API server to complete deletion.
```
Something worth to note here is that this will require 2x API calls from kubelet -> apiserver for these types of pods.

If before kubelet was simply deleting the pod: only one HTTP call was needed (i.e. HTTP DELETE), now this will require two HTTP calls, i.e. (1) Patch the pod status with phase (2) Delete the pod.

I assume the cases of these types of pods is relatively rare, so I don't think is a large concern, but worth to document.

mimowo replied:
I think (please correct me if I'm missing something), that there is one additional API call needed kubelet -> API server to PATCH the pod (to move the transition phase). 

The new flow of requests in this scenario:
1. a scheduled pod (with nodeName != nil) is in Pending phase (same)
2. DELETE request to API server is sent by user or another k8s component (e.g. scheduler) -> the pod is terminating
3. [new] PATCH by Kubelet to transition the Pod into Failed phase
4. PATCH the pod to remove the finalizer by the job controller
5. Kubelet clears up the resources as there is no Pod

I'm not aware of Kubelet sending DELETE request after the PATCH (3.). Let me know if this is the case and where I can find it in code. 

Anyway, we will have one additional PATCH kubelet -> api server, but indeed the situation should be rare as it requires DELETE which needs to be send by a user or another component for the Pod to be terminating.

mimowo replied:
Update: added a paragraph on this. PTAL.

bobbypage replied:
Kubelet will need to send the final delete request to the api-server to actually force delete the pod with zero grace period seconds (just like it does today). So today the flow is:

1. Pod created with spec.nodeName != nil
2. Scheduler sets spec.nodeName
3. Pod is pending
4. API-server delete recieves delete with some deletion grace period seconds > 0; api-server set's a deletionTimestamp on the pod
5. Kubelet sends HTTP DELETE to api-server with zero grace period seconds to remove the pod from the api-server

With this feature we need an additional call between steps 4 and 5 to send the HTTP patch to update the phase to failed.

mimowo replied:
I see, I've checked that it is send from here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L745-L758.  I guess this is to speed up the deletion in API server, because otherwise API server would wait until the original grace period elapsed?

Still it holds that there is a single additional PATCH request.

alculquicondor replied:
There is a PRR question about additional API calls. Please update.

mimowo replied:
Added a small note there.

bobbypage replied:
> I see, I've checked that it is send from here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L745-L758. I guess this is to speed up the deletion in API server, because otherwise API server would wait until the original grace period elapsed?

That is the final delete sent to the API server from the kubelet to delete the pod from etcd which is what is expected in the normal case. Only the kubelet knows when it is safe to force delete the pod (since it should only force delete the pod once the pod is terminal).

bobbypage commented:
I don't see any large issues with proposal here, but I suggested a few updates to the wording in the KEP to be a bit more precise on the changes here.

bobbypage commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -883,20 +889,51 @@ is disabled [Disconnected node when taint-manager is disabled](#disconnected-nod
+[Mark Pending Terminating pods as Failed by Kubelet](https://github.com/kubernetes/kubernetes/pull/115331).
+
+Also, for the second iteration of Beta (1.27) we are going to expand the
+feature documentation to explain this change.
```
It may be worth to note here some of the cases that pod may be stuck in pending. In general for a pod to go into pending all the containers have to been started. So if a container does not start, for example the init containers did not succeeded or one of the containers could not start, it will remain in pending. xref: https://github.com/kubernetes/kubernetes/blob/d9f21e55df4cd95f5e6c9ea1d7109a2a767920fb/pkg/kubelet/kubelet_pods.go#L1408-L1418

mimowo replied:
Thanks, updated the point about documentation, to provide the list of example scenarios.

soltysh reviewed (approved):
Minor nit, but overall this lgtm from sig-apps pov
/approve

leaving lgtm to Aldo

alculquicondor commented:
LGTM, but I think the official one belongs to SIG Node @SergeyKanzhelev

bobbypage commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -871,32 +878,110 @@ not added to the pod for a long enough time (for example 2 minutes).
+make progress after retrying. Once such a Pod is also terminating, Kubelet is not
+going to attempt starting of the pod.
+
+##### Implementation progress and plan
+
```
A couple items I think we should discuss in the KEP:

1. When the kubelet's sets the phase to failed for these pods, what should the message be communicated on pod status? i.e. phase=Failed, but what about Status and Message [1]? Ideally it would contain some info of why the pod was stuck in pending to help users debug.
2. We should ideally ensure that the kubelet reports the latest `ContainerStatuses` on the pod as well as part of this last update. If not, we have some bug in kubelet we need to fix.
3. I chatted about this change briefly with @smarterclayton. Clayton mentioned it is arguably a bug that this not happen today, i.e. pods are expected to go through their entire lifecycle and end at Succeeded or Failed. So it might be worth to call out in the KEP, that one of the reasons for this change is we want to ensure that all pods will end up in terminal state, even if they did not start (i.e. were stuck in pending).


[1] https://github.com/kubernetes/kubernetes/blob/49ddc510bc6b57fa430f779372b71144d328995c/staging/src/k8s.io/api/core/v1/types.go#L4023-L4027

mimowo replied:
Thanks for the comment, I will update the KEP with a review of a couple of scenarios leading to the state. 
However, answering the comment here:
[1] we may need to end up with a generic messages such as {Reason: Evicted, Message: Evicted as pending and terminating}. The reason is that the Pending state would not necessarily make it stuck by itself. The reason for the DELETE request should be added in a pod condition by an external controller. For example, scheduler, will add `DisruptionTarget` condition with a message saying that it is preempting the pod. Still, the pod could get stuck in `Pending` phase, even without the DELETE, for example due to invalid image reference, but I think such scenarios deserve a dedicated KEP, because it might be hard to discriminate between cases where the image is misconfigured or there are transient connectivity issues.
[2] I'm intending to use the same scheme for evicting pods as currently used for example for pods which exceed the activeDeadlineSeconds timeout. This reports the `containerStatuses`.
[3] I agree with this, maybe it could be handled as a bugfix outside of the KEP (it is needed for the KEP, but maybe something else). Still, the KEP creates a nice communication channel to align on the approach of fixing it . Once done, we could consider cherry-picking it for 1.26.

alculquicondor replied:
I think the reason should be `Deleted` (something deleted the pod). And the message could be "Deleted while Pending". But..... can we leave this discussion to code review? It sounds minor.

mimowo replied:
@bobbypage updated the KEP with suggestions for [1] and [3] and provided a short review of example scenarios leading to the state. PTAL. As for [2] I think this could be an implementation thing.

bobbypage replied:
Thank you for all the updates in the KEP lgtm!

alculquicondor commented:
/assign @johnbelamaric
for update to Prod Readiness questionnaire

mimowo commented:
@bobbypage @SergeyKanzhelev I've updated the PR, addressing the comments. Please complete the review or let me know if there is something more requiring an update.

johnbelamaric commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -1815,7 +2109,10 @@ previous answers based on experience in the field.
 ###### Will enabling / using this feature result in any new API calls?
 
-Yes. An API call to append a Pod condition when deleting the Pod.
+Yes. An PATCH API call to append a Pod condition when deleting the Pod. Also,
```
nit: s/An PATCH/A PATCH/

Should not hold up the PR.

mimowo replied:
Done

johnbelamaric commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -1815,7 +2109,10 @@ previous answers based on experience in the field.
+Yes. An PATCH API call to append a Pod condition when deleting the Pod. Also,
+one PATCH API call to set Pod's phase as `Failed` in scenarios (2nd and 3rd)
+described under [Marking pods as Failed](#marking-pods-as-failed). Note that,
+in the 1st scenario the the phase is set along with adding the Pod condition.
```
nit: s/the the/the/

mimowo replied:
Done

johnbelamaric reviewed (commented):
Thanks, this looks good from a PRR perspective, except we have added one question to the PRR this cycle:

https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md?plain=1#L723

Can you answer that please?

bobbypage commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -871,32 +880,314 @@ not added to the pod for a long enough time (for example 2 minutes).
+
+Note that, as PodGC sends DELETE request for such pod (DELETE request was sent
+prior to this KEP, but without setting the phase to `Failed`), it becomes also
+terminating (has `deletionTimestamp` set).
```
nit: I would avoid using the term "terminating", as for the pod to be terminating in the traditional sense, the pod has have had started. If the pod never started, it may not even be terminating. 

I think the term we should use here is "deleting". See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/object-lifecycle.md#states here it is defined. (and you can maybe link to it from the KEP)

mimowo replied:
Yeah, I guess I was influenced by the terminology used in kubectl, there `Terminating` is displayed even if it has not started. 

For example, if I create a pod via the following Job spec (unschedulable):
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: job-unchedulable
spec:
  template:
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: kubernetes.io/os
                operator: In
                values:
                - something
      restartPolicy: Never
      containers:
      - name: job-unchedulable
        image: centos:7
        command: ["bash"]
        args: ["-c", 'echo "Hello world"']
        imagePullPolicy: IfNotPresent
  backoffLimit: 0
```
Then, do `kubectl delete`, it shows up as `Terminating` in the console for `kubectl get pods`. 

Also, the `Terminating` word is used in PodGC code (prior to this work): https://github.com/kubernetes/kubernetes/blob/8e20effbb7b99c9c5caad4b8f947ca645bbaf21b/pkg/controller/podgc/gc_controller.go#L133.

mimowo replied:
For now, I reference the link and add a note that this is an alternative name. WDYT?

dchen1107 replied:
SGTM for now.

bobbypage replied:
Sounds good, thanks for adding the reference. 

Just as fyi, the term terminating in kubectl comes from [here](https://github.com/kubernetes/kubernetes/blob/0f37b3120643580f632ca12d3e174e7ec447948c/pkg/printers/internalversion/printers.go#L853-L855). `Terminating` shows up if deletion timestamp is set.

My comment is just that in my mind there is distinction between deleting and terminating. `Deleting` == deletion timestamp is set, terminating == the kubelet is terminating the pod, i.e. terminating containers, cleaning up resources etc. As described in the KEP, in any case the pod is `Deleting`, the phase should be updated to terminal.

bobbypage commented on keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md:
```diff
@@ -871,32 +880,314 @@ not added to the pod for a long enough time (for example 2 minutes).
+
+**Correct config, but image takes long to download, pod deleted by a user in the meanwhile**
+
+This scenario is a little bit different, the config is correct, but the
+pod is deleted by a user while in the `Pending` phase. In that case, the
+pods transition into the `Running` phase and fail soon after. With the proposed
+change the transition will happen earlier, thus saving resources.
+
```
nit: I'm not super clear why this scenario will result in phase=Failed being applied, but not the previous scenario. Not needed for the KEP doc itself, but I think we should understand what is going on here...

mimowo replied:
In this scenario the pod actually transitions to the `Running` phase, but the container is killed, respending the grace period. If the container completes with exitCode `0` within the graceful period it actually ends the entire pod in Succeed state, otherwise it ends with exitCode `137`.

bobbypage replied:
gotcha, thanks for clarifying this in the doc.

bobbypage commented:
I have one more comment about the terminology used in the KEP (terminating vs deleting), but LGTM on the content, thank you for all of the updates and clarifying the scenarios.

mimowo commented:
> Thanks, this looks good from a PRR perspective, except we have added one question to the PRR this cycle:
> https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md?plain=1#L723
> Can you answer that please?

Done, PTAL. The feature only introduces an additional API PATCH call, what is answered in the other point.

johnbelamaric commented:
Ok, PRR looks good and is approved (no prow command needed this time, but want to make it explicit for the enhancements team).

bobbypage commented:
Thanks for all the updates clarifications in the KEP. 

/lgtm

dchen1107 commented:
Thanks @bobbypage 

I also reviewed it. 

/lgtm
/approve from SIG Node perspective.

mimowo commented:
@bobbypage @dchen1107 thank you for completing the review!
